src/ctia/entity/vulnerability.clj

```diff
@@ -109,29 +115,69 @@
 (def vulnerability-enumerable-fields
   [:source])
 
+(s/defn search-by-cpe-ids [{{:keys [get-store]} :StoreService
+                            :as services} :- APIHandlerServices]
+  (context "/cpe_ids" []
+    (let [capabilities :search-vulnerability]
+      (GET "/" []
+        :return PartialVulnerabilityList
+        :query [{cpe-ids-string :cpe_ids :as params}
+                (st/merge VulnerabilitySearchParams {:cpe_ids (swagger-schema/describe s/Str "CPE ids to search by")})]
+        :summary "List Vulnerabilities by CPE ids"
+        :description (routes.common/capabilities->description capabilities)
+        :capabilities capabilities
+        :auth-identity identity
+        :identity-map identity-map
+        (let [cpe-ids (str/split cpe-ids-string #"\s+")
+              vulnerability-ids
+              (loop [query-params params
```

Comment:
This loop is exactly what does `list-all-pages` and the reason of OOM. If we do this we do not control anymore the volume of data loaded in memory in the server for this request. 
You need to expose pagination parameters and let the client paginate.

Reply:
Thank you for looking at this. Please consider the following and let me know your thoughts:

- There are two queries being used within this endpoint 
- The first query which is within the loop, is using server side paging, and processing each page individually
- The purpose of the loop, is to construct a vector of IDs which are used to construct the second query.
- This is different from `list-all-pages` in that each page is being processed individually by `cpe/vulnerabilities->ids` to extract only the ids of vulnerabilities that have matching configurations.
- Unlike `list-all-pages` which accumulates all of the query responses into memory, this loop only accumulating the `ids` of vulnerabilities with matching configurations.
- It is frequently the case that the first query in the loop will not find any vulnerability ids with matching configurations in the first several pages. Were we to use client side paging in the first query, this could cause the second query to incorrectly not return anything to the user.
- The second query employs client side paging

Reply:
I understand your point of view and made my previous remark with this understanding of the algorithm. The issue is to ensure the protection of the server over potentially unexpected cases (e.g. unexpected input, unexpected state of the vulnerability store, high number of concurrent queries).
I propose that you use a [lazyseq](https://clojuredocs.org/clojure.core/lazy-seq) like [here](https://github.com/threatgrid/ctia/blob/master/src/ctia/task/migration/migrate_es_stores.clj#L126). 
It would be a simple replacement of your `loop`:
```clojure

(defn list-vulnerability-ids
    [query-params]
    (lazy-seq
	 (let [{:keys [data paging]}
                 (ctia-store/list-records (get-store :vulnerability)
                                               {:one-of {:configurations.nodes.cpe_match.cpe23Uri cpe-ids
                                                         :configurations.nodes.children.cpe_match.cpe23Uri cpe-ids}}
                                               identity-map
                                               (merge query-params {:fields [:configurations.nodes :id]}))
                      ids (cpe/vulnerabilities->ids cpe-ids data)]
                  (if-let [next-params (:next paging)]
                     (concat ids (list-vulnerability-ids next-params))
                     ids))))
```
It should be tested, but the pagination should be done with the parameters of this function, right?
Since it's a little bit tricky, you should carefully check this pagination.

Reply:
I do like the direction you are going in with a lazy sequence of ids. Would the proposed implementation using `concat` be vulnerable to [Clojure Don’ts: Concat](https://stuartsierra.com/2015/04/26/clojure-donts-concat)?

Reply:
> Would the proposed implementation using `concat` be vulnerable to [Clojure Don’ts: Concat](https://stuartsierra.com/2015/04/26/clojure-donts-concat)?

That was a quick proposition, but I think that `into` would be more appropriate. There are many traps that force the realization of a lazy collection, that's why you should properly test this lazyness.
I propose to invoke some clojure god and ask if @frenchy64 could help us to properly do it :-).

Reply:
FWIW I think lazy-seq + concat is safe. I believe the blog is talking about loop (or any strict op) + concat. IIUC the stackoverflow happens when you call `first` on a large stack of `concat`s. I don't think this is possible with `lazy-seq`, but is easy to reproduce with reduce or loop.